src/bin/pg_autoctl/pgctl.c

```diff
@@ -975,57 +981,80 @@ pg_write_recovery_conf(const char *pgdata,
 					   const char *replicationSlotName)
 {
-	char recoveryConf[BUFSIZE] = { 0 };
-	char *recoveryConfEnd = recoveryConf;
 	char recoveryConfPath[MAXPGPATH];
-	int recoveryConfLength = 0;
+	PQExpBuffer content = NULL;
 
 	log_trace("pg_write_recovery_conf");
 
 	/* build the contents of recovery.conf */
-	recoveryConfEnd += sprintf(recoveryConfEnd, "standby_mode = 'on'");
-	recoveryConfEnd += sprintf(recoveryConfEnd,
-							   "\nprimary_conninfo = %s", primaryConnInfo);
-	recoveryConfEnd += sprintf(recoveryConfEnd,
-							   "\nprimary_slot_name = '%s'", replicationSlotName);
-	recoveryConfEnd += sprintf(recoveryConfEnd,
-							   "\nrecovery_target_timeline = 'latest'");
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\n");
+	content = createPQExpBuffer();
+	appendPQExpBuffer(content, "standby_mode = 'on'");
+	appendPQExpBuffer(content, "\nprimary_conninfo = %s", primaryConnInfo);
+	appendPQExpBuffer(content, "\nprimary_slot_name = '%s'", replicationSlotName);
+	appendPQExpBuffer(content, "\nrecovery_target_timeline = 'latest'");
+	appendPQExpBuffer(content, "\n");
 
-	recoveryConfLength = recoveryConfEnd - recoveryConf;
+	/* memory allocation could have failed while building string */
+	if (PQExpBufferBroken(content))
+	{
+		log_error("Failed to allocate memory");
+		destroyPQExpBuffer(content);
+		return false;
+	}
 
 	join_path_components(recoveryConfPath, pgdata, "recovery.conf");
 
 	log_info("Writing recovery configuration to \"%s\"", recoveryConfPath);
 
-	if (!write_file(recoveryConf, recoveryConfLength, recoveryConfPath))
+	if (!write_file(content->data, content->len, recoveryConfPath))
 	{
 		/* write_file logs I/O error */
+		destroyPQExpBuffer(content);
 		return false;
 	}
 
+	destroyPQExpBuffer(content);
 	return true;
 }
 
 
 /*
- * escape_recovery_conf_string escapes a string that is used in a recovery.conf file
- * by converting single quotes into two single quotes.
+ * escape_recovery_conf_string escapes a string that is used in a recovery.conf
+ * file by converting single quotes into two single quotes.
  *
  * The result is written to destination and the length of the result.
  */
-static int
-escape_recovery_conf_string(char *destination, const char *recoveryConfString)
+static bool
+escape_recovery_conf_string(char *destination, int destinationSize,
+							const char *recoveryConfString)
 {
 	int charIndex = 0;
 	int length = strlen(recoveryConfString);
 	int escapedStringLength = 0;
 
+	/* we are going to add at least 3 chars: two quotes and a NUL character */
+	if (destinationSize < (length+3))
+	{
+		log_error("BUG: failed to escape recovery parameter value \"%s\" "
+				  "of %d bytes in a buffer of %d bytes",
+				  recoveryConfString, length, destinationSize);
+		return false;
+	}
+
 	destination[escapedStringLength++] = '\'';
 
 	for (charIndex = 0; charIndex < length; charIndex++)
 	{
 		char currentChar = recoveryConfString[charIndex];
+
+		if (destinationSize < (charIndex+1))
```

Comment:
we should compare against escapedStringLength

Reply:
this was we could think about removing the previous `(destinationSize < (length + 3))` check

Reply:
I improved the error checking here, thanks to your review. Please have another look?